README.md
# AngularJS Weather Single Page App (SPA) from Anthony Alicea's Udemy Angular course.
## - with modifications -

This is an [Angular.js](https://angularjs.org/) framework SPA taken from [Anthony Alicea's](http://www.anthonypalicea.com) course on [Angular at Udemy](https://www.udemy.com/learn-angularjs/learn/#/). It pulls a weather data object returned in JSON from [Open Weather Map](http://openweathermap.org/).

I have made a number of modification to it, using Bootstrap classes and the [Angular ngClass directive](https://docs.angularjs.org/api/ng/directive/ngClass).

* I have added in a function in the controllers.js, accessed in the weatherReport.htm directive template that computes the temperature in both ℉ (Farenheit) and ℃ (Celsius).
* I also pull the weather objects main & description in weatherReport.htm from the returned weather object and put those in each of the results.
* I added two rows of Bootstrap buttons to choose how many days forecast you are viewing for the city's weather forecast in forecast.htm, with ng-class substituting out correctly with the 'btn-primary' or 'btn-diabled' depending if the button has been chosen or not.

For my education and edification purposes only, not for sale and giving credit for original code to [Anthony Alicea's](http://www.anthonypalicea.com).


CHANGELOG.md

* [view commit](https://github.com/faddah/angular-weather-spa/commit/1ddf66b) &bull; per instructions in udemy training module, separated out weatherResult and weatherAPI into a separate service in services.js, then leaned up the function in controller.js, including how it is called in home.htm and index.htm. work, runs. all go.
* [view commit](https://github.com/faddah/angular-weather-spa/commit/301417e) &bull; change to README.md to add link for ngClass doc on angular.js web site.
* [view commit](https://github.com/faddah/angular-weather-spa/commit/34a8f16) &bull; change to package.json
* [view commit](https://github.com/faddah/angular-weather-spa/commit/c88ef94) &bull; added README.me describing project and crediting with links Anthony Alicea and Udemy, mentioning my additions. prepping to send to github remote repository.
* [view commit](https://github.com/faddah/angular-weather-spa/commit/87e66f5) &bull; added a package.json with npm init, added links in heading of index.html to link to angularjs.org and openweathermap.org.
* [view commit](https://github.com/faddah/angular-weather-spa/commit/a15276e) &bull; everything working! errors were passing in forecast.html 'date' as ng-repeat var instead of 'w', as it was in every other place, now data pulls fine from openweathermap.org and thanks to folks helping me with ngClass on stackoverlow, ng-class in forecast.htm now exchanges out correct bootstrap class for buttons depending if button has been selected ('disabled') or not. custom directive in angular created for ng-repeat of weather listing, routes, services, controlle
 and directives all separated out into their correct, respecitve js files. considering making this into a a phonegap\cordova app for iPhone and android.
* [view commit](https://github.com/faddah/angular-weather-spa/commit/0a2cfad) &bull; commit after downloading files from dropbox after sending them up there after working on them on chromebook. the city transferring properly in from home page to forecast page is still not happening, investigating why, hopefully fixed by next iteration.
* [view commit](https://github.com/faddah/angular-weather-spa/commit/22d1432) &bull; got date formatted correctly in forecast result list by using  settings for date bindings. still can't get city to be anything but Portland, don't know what broke. will investigate and hopefully fix by next iteration.
* [view commit](https://github.com/faddah/angular-weather-spa/commit/37b6b1e) &bull; about one-third to one-half way through the module - got weather & weather description to show up now by remembering to use the dad-blamed weather[0].description... etc., because in the returned JSON object it's in an array - oh joy. who is the idiot who keeps thinking that putting nested arrays in JSON files\/objects is a good idea? because they are idiots. let objects be freakin' objects, easier to access all members that way. anyways, still need to convert date and get routes requesting more or less weather days running proper. next iteration, promise.
* [view commit](https://github.com/faddah/angular-weather-spa/commit/3c6696b) &bull; half-way through modules - added appid also for accessing the API key for openweathermap.org and have temperature correctly displaying now, converted from Kelvin to Farenheit; still can't get weather and description data members to show, working on that, and need to get routes working for more days to show and convert date.
* [view commit](https://github.com/faddah/angular-weather-spa/commit/0b86819) &bull; first, initial commit after downloading untouched files for this module from Udemy.</li>


controllers.js
// CONTROLLERS

weatherApp.controller('homeController', ['$scope', '$log', '$routeParams', '$location', 'cityService', function($scope, $log, $routeParams, $location, cityService) {

  $scope.city = cityService.city;

  $scope.$watch('city', function() {
    cityService.city = $scope.city;
  });

/* * * Former method to go to forecast.htm template from home.htm after submitting city. now being handled by $scope.submit() function,
need to add $window service to controller injected dependencies if you use this, otherwise use $location with $scope.submit instead.
  $scope.goToForecast = function() {
    // $location.path('#/forecast');
    // $location.replace();
    $window.location.href = '#/forecast';
  };

  * * * * * * * * * */

  $scope.submit = function() {
    $location.path("/forecast");
  };

}]);

weatherApp.controller('forecastController', ['$scope', '$log', '$routeParams', 'cityService', 'weatherService', function($scope, $log, $routeParams, cityService, weatherService) {

  $scope.city = cityService.city;

  $scope.days = $routeParams.days || '1';

  $scope.weatherResult = weatherService.GetWeather($scope.city, $scope.days);

  // $log.info($scope.weatherResult);

  $scope.convertToFarenheit = function(degK) {

    return Math.round((degK* 1.8) - 459.67);

  };

  $scope.convertToCelsius = function(degK) {

    return Math.round(degK - 273.15);

  };

  $scope.convertToDate = function(dt) {

    return new Date(dt * 1000);

  };

}]);


app.js
// MODULE

var weatherApp = angular.module('weatherApp', ['ngRoute', 'ngResource']);

// ROUTES

weatherApp.config(function($routeProvider) {
  
  $routeProvider
  
  .when('/', {
    templateUrl: 'pages/home.htm',
    controller: 'homeController'
  })
  .when('/forecast', {
    templateUrl: 'pages/forecast.htm',
    controller: 'forecastController'
  });
  
});

// SERVICES

weatherApp.service('cityService', function() {
  
  this.city = 'Portland, OR';
  
});

// CONTROLLERS

weatherApp.controller('homeController', ['$scope', '$log', '$routeParams', 'cityService', function($scope, $log, cityService) {
  
  $scope.city = cityService.city;
  
  $scope.$watch('city', function() {
    cityService.city = $scope.city;
  });
  
}]);

weatherApp.controller('forecastController', ['$scope', '$log', '$resource', '$routeParams', 'cityService', function($scope, $log, $resource, $routeParams, cityService) {
  
  $scope.city = cityService.city;
  
  $scope.days = $routeParams.days || 2;
  
  $scope.weatherAPI = $resource("http://api.openweathermap.org/data/2.5/forecast/daily", { callback: "JSON_CALLBACK" }, { get: { method: "JSONP" } } );
  
  $scope.weatherAPI.appid = 'f857b7ab2bf18470438f14cd1da0781f';
  
  $scope.weatherResult = $scope.weatherAPI.get({ appid: $scope.weatherAPI.appid, q: $scope.city, cnt: $scope.days });
  
  // $log.info($scope.weatherResult);
  
  $scope.convertToFarenheit = function(degK) {
    
    return Math.round((degK* 1.8) - 459.67);
    
  };
  
  $scope.convertToCelsius = function(degK) {
    
    return Math.round(degK - 273.15);
    
  };
  
  $scope.convertToDate = function(dt) {
    
    return new Date(dt * 1000);
    
  };
  
}]);